src/pid-file/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2020 Andrej Shadura.
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// This file is derived from the bsd-pidfile-rs project, available at
+// https://github.com/andrewshadura/bsd-pidfile-rs. It was incorporated
+// directly into Materialize on August 12, 2020.
+//
+// The original source code is subject to the terms of the MIT license, a copy
+// of which can be found in the LICENSE file at the root of this repository.
+
+//! PID file management for daemons.
+//!
+//! The `pid-file` crate wraps the `pidfile` family of functions provided by BSD
+//! systems that provide mutual exclusion for daemons via PID files.
+//!
+//! Much of the code is inherited from [pidfile_rs], but the build system
+//! bundles the necessary subset of libbsd rather than relying on libbsd to be
+//! installed.
+//!
+//! [pidfile_rs]: https://docs.rs/pidfile_rs
+
+#![warn(missing_docs)]
+
+use std::convert::TryInto;
+use std::ffi::{CString, NulError};
+use std::fmt;
+use std::fs::Permissions;
+use std::io;
+use std::os::unix::ffi::OsStrExt;
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+use std::ptr;
+
+use ore::option::OptionExt;
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct pidfh {
+    private: [u8; 0],
```

Comment:
Is this like the rust equivalent of flexible array members or some such - a way to say "this is some arbitrary pile of bytes allocated/managed on the other side of the ffi boundary, don't worry about it"?

Reply:
Yep, exactly. This SO post has a good explanation of the situation: https://stackoverflow.com/a/38315613/1122351

tl;dr an RFC for a better way was proposed and accepted, but naturally it's stuck in committee: rust-lang/rust#43467.